Core/Quests: implement Quest Greeting (DB table quest_greeting)

ForesterDev opened the pull request:
This is a partial port of https://github.com/TrinityCore/TrinityCore/commit/22e230ecec1c7ce4eb4daeb60adfa03098271f34

<!--- (**********************************)
      (** Fill in the following fields **)
      (**********************************) --->

**Changes proposed:**

-  Partial port of https://github.com/TrinityCore/TrinityCore/commit/22e230ecec1c7ce4eb4daeb60adfa03098271f34 into 3.3.5 to enable use of Quest Greetings

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Updates #19756  (insert issue tracker number)


**Tests performed:** (Does it build, tested in-game, etc.)
Tested

**Known issues and TODO list:** (add/remove lines as needed)

- [x] Implement locales for quest greetings

<!--- Notes
- Enable the setting "[√] Allow edits from maintainers." when creating your pull request.
- If this PR only contains SQL files, open a new issue instead and post or link the SQL in the issue.
- When adding new SQL files, name them 9999_99_99_99_db_name.sql to reduce the chance of possible merge conflicts.
--->

Commit:
Core/Quests: implement Quest Greetings. Partial port https://github.com/TrinityCore/TrinityCore/commit/22e230ecec1c7ce4eb4daeb60adfa03098271f34

Commit:
fix build

Shauren commented:
Why is this only a partial port?

ForesterDev commented:
I don't port some parts of commit, e.g part with npc_text, because not all npc_texts data have BroadcastTextIds. I also leave old QEmote and Title behaviour for missing greetings data

Commit:
add log error for unknown type

Commit:
add check for greetEmoteType

tkrokli commented:
First of all, thanks for starting this job. (I was too slow to get it started myself.) 👍

Secondly, here is what I think will be needed for the RBAC part of the PR (not sure if you have planned it already):

Add the file `sql/updates/auth/2017_06_06_99_auth.sql` (or use dates/numbers you prefer) to your local source:
```sql
DELETE FROM `rbac_permissions` WHERE `id`= 843; -- this id does not exist in TDB 335.63
INSERT INTO `rbac_permissions` (`id`,`name`) VALUES (843, 'Command: reload quest_greeting');
```
After this, you will need to run the auto DB updater (worldserver) and add the contents from the last line in your local 3.3.5 `auth.updates` table to the end of **line 489**
```sql
INSERT INTO `updates` VALUES ('2014_11_10_00_auth.sql','0E3CB119442D09DD88E967015319BBC8DAFBBFE0','ARCHIVED','2015-03-21 21:44:12',0), ...
```
in the file `sql/base/auth_database.sql` in your PR branch.

You will also need to update **line 402**
```sql
INSERT INTO `rbac_permissions` VALUES ( [...] )
```
with `(843,'Command: reload quest_greeting'),` between `(841,'Command: .debug neargraveyard'),` and `(852,'Command: go offset'),`

You will also need to update **line 378**
```sql
INSERT INTO `rbac_linked_permissions` VALUES (192,21),(192,42),(192,43),(192,193), [etc.]
```
with `(196,843)` between `(196,841),` and `(197,232),`
***
After all these tasks, push your changes to this PR (your branch `ForesterDev:quest_greetings` )
BTW, the **line numbers in bold** are based on current 3.3.5 TC (TDB 335.63 + updates).

Commit:
update RBAC

ForesterDev commented:
@tkrokli , done

tkrokli commented:
Thank you for the update. 👍  Looks good to me.

To the developer who will merge it: This PR does **_not_** close issue #19756 , only updates it, because there are several SQL queries in that issue which will be added to the table once the table is part of the 3.3.5 source.

offl commented:
You can add SQL part to this PR.

Aokromes committed:
Rename 2017_06_05_99_world.sql to 2017_06_05_99_world_335.sql

ForesterDev commented:
This PR is WIP now. I want to implement locales before merging to TC

Commit:
implement locales for quest_greeting table

ForesterDev commented:
If someone has verified db data, i can add it to this PR

Commit:
forgotten changes for last commit

offl commented:
What do you mean by "verified"? You mean sniffed data? All we have is the data from movies.
```sql
DELETE FROM `quest_greeting` WHERE `ID` IN (234, 235, 237, 238, 241, 261, 714, 823, 2519, 2930, 3391, 3419, 3616, 3649, 4077, 6171);
INSERT INTO `quest_greeting` (`ID`, `Type`, `GreetEmoteType`, `GreetEmoteDelay`, `Greeting`, `VerifiedBuild`) VALUES
(234, 0, 1, 0, "A foul corruption has crept into Westfall.  While I was upholding my duty on the battlefields of Lordaeron these honest farms were overrun and made into hide-outs for thugs and murderers. The People's Militia needs your help.", 0),
(235, 0, 0, 0, "Welcome to our humble abode! It's always nice to see a friendly face. And what strong arms you have. My husband and I are always looking for help around the farm. Now that most the good folk have left, it's hard to find an able body to help out.", 0),
(237, 0, 0, 0, "Nothing but trouble in these parts.  I tried to tell that fool Saldean to get out while he still could but he won't hear of it.  But I ain't no fool.  Verna and I are gonna mosey on out as soon as we get this wagon fixed.", 0),
(238, 0, 0, 0, "Sometimes I think there's a big gray cloud in the sky, just raining down bad luck upon us. First, we're driven off our land, and now we can't even get out of Westfall. Everything's a mess. Something needs to be done.", 0),
(241, 0, 0, 0, "Hey there, friend.  My name's Remy.$BI'm from Redridge to the east, and came here looking for business, looking for business.  You got any...got any??", 0),
(261, 0, 0, 0, "Hello, citizen.  You look like a $g man : woman; with a purpose - do you have business with the Stormwind Army?", 0),
(714, 0, 0, 0, "Greetings, $c! Fine day for hunting, wouldn't you say? I've been having more than a little luck with boars, myself. Perhaps you'd like a shot?", 0),
(823, 0, 1, 0, "Hello there, $c.  Normally I'd be out on the beat looking after the folk of Stormwind, but a lot of the Stormwind guards are fighting in the other lands.  So here I am, deputized and offering bounties when I'd rather be on patrol...", 0),
(2519, 0, 0, 0, "Hello! The spirits say you here to aid my chief. Say it loud, they do.", 0),
(2930, 0, 0, 0, "Corruption sneaks into nature's grove. The forest must be protected at all costs.", 0),
(3391, 0, 0, 0, "Thrall paid me and my boys well for helping out with the construction of Orgrimmar, so I decided to set up a port here. We do most of our business through Booty Bay and Baron Revilgaz.", 0),
(3419, 0, 1, 0, "For our people to survive, we must study the ways of nature and learn its secrets.", 0),
(3616, 0, 0, 0, "The wind whispers to those who listen...$B$BDo you hear it?", 0),
(3649, 0, 1, 0, "Dark forces encroach upon our borders, ancient taints resurface, and new evils emerge to topple the delicate balance of the land. In times of such darkness, we all must be vigilant.", 0),
(4077, 0, 0, 0, "If I can get the proper ingredients, we can create some explosives that will not only allow me to damage the Venture's Co.'s operations, but also cause a distraction.$B$BI know you're impressed, but it gets better... I'll tell you more soon enough.", 0),
(6171, 0, 1, 0, "You will be tested many times by the darkness that surrounds our lands, $N. But to ensure you are always prepared, we will often ask tasks of you so you are at the height of your own power.$B$BGo with the Light, and remain virtuous.", 0);
```

Commit:
add some db data

Aokromes commented:
better leave db data apart, because the new db structure cannot be added into master and adding new db structure+db data on same commit will mess things.

tkrokli commented on sql/updates/world/3.3.5/2017_06_05_99_world_335.sql:
```diff
@@ -0,0 +1,47 @@
+INSERT INTO `command` (`name`, `permission`, `help`) VALUES
+('reload quest_greeting', 843, 'Syntax: .reload quest_greeting\nReload quest_greeting table.');
+
+DELETE FROM `command` WHERE `name`='reload quest_greeting';
```
I hope this is a copy-paste mistake, because it deletes the previous insert. Maybe you meant
```sql
DELETE FROM `command` WHERE `name`='reload quest_greeting_locale';
```

Commit:
fix copy-paste typo and remove db data

Commit:
add log

funjoker commented:
broadcast text? localiuation ?

ForesterDev commented:
There is no broadcast texts for this strings

offl commented:
There is no broadcast_text for quest_greeting. So we can't add localization.

Currently we can only sniff localization for NPC 3391, because the text never was changed or removed.

funjoker commented:
Every string which is seen ingame has/needs localization.
If there are no strings inside broadcast_text may add quest_greeting_locale

tkrokli commented:
`src/server/game/Accounts/RBAC.h` :
```cpp
    RBAC_PERM_COMMAND_NPC_SPAWNGROUP                         = 856, // reserved for dynamic_spawning
    RBAC_PERM_COMMAND_NPC_DESPAWNGROUP                       = 857, // reserved for dynamic_spawning
    RBAC_PERM_COMMAND_GOBJECT_SPAWNGROUP                     = 858, // reserved for dynamic_spawning
    RBAC_PERM_COMMAND_GOBJECT_DESPAWNGROUP                   = 859, // reserved for dynamic_spawning
    RBAC_PERM_COMMAND_LIST_RESPAWNS                          = 860, // reserved for dynamic_spawning
```
`    RBAC_PERM_COMMAND_RELOAD_QUEST_GREETING_LOCALE           = 857,`
needs to be moved to at least 861.

Notice also that this enum is not yet implemented in `master` branch.

Aokromes commented:
Conflicting files
sql/base/auth_database.sql
src/server/game/Accounts/RBAC.h
src/server/game/Entities/Creature/GossipDef.cpp
src/server/game/Globals/ObjectMgr.h
src/server/scripts/Commands/cs_reload.cpp

Commit:
Merge branch '3.3.5' into quest_greetings

Commit:
fix sql

Commit:
fix codestyle

Commit:
fix build

Aokromes commented:
Conflicting files
sql/base/auth_database.sql

Commit:
Resolve conflicts in auth_database.sql

- rename files to newer date than existing files
- add missing changes from 3.3.5 branch

Commit:
Merge branch '3.3.5' of https://github.com/TrinityCore/TrinityCore into quest_greetings

# Conflicts:
#	sql/base/auth_database.sql

brotalnia commented:
I believe gossip text is not supposed to be shown when there is no quest greeting.

Look at this video:
https://youtu.be/558vNOJ7rvA?t=1m20s

There is not text at all, not even Greetings $n.

tkrokli commented:
Well, that is partially right. It depends on if the NPC has gossip or not (based on sniff data of the NPC). It is correct when the NPC does not have a fixed gossip text. (Maybe some NPCs need to be rechecked?) Remember also that `quest_greeting` is shown only for some of the NPCs using gossip, not all of them, and those who use `quest_greeting` only show it after some specific quests or based on some condition.

Aokromes committed:
Update RBAC.h

tkrokli commented on sql/updates/auth/3.3.5/2017_07_30_00_auth.sql:
```diff
@@ -0,0 +1,15 @@
+INSERT INTO `rbac_permissions` (`id`,`name`) VALUES
+(843, 'Command: reload quest_greeting');
+
+DELETE FROM `rbac_linked_permissions` WHERE `id` = 843;
```
Wrong field name, needs to be `linkedId` (see below).

Commit:
Merge branch '3.3.5' into quest_greetings

Commit:
Update and rename 2017_07_30_00_auth.sql to 2017_10_08_00_auth.sql

- 843 is `linkedId` , not `id` in this case.
- updated sql file name to match the `updates` table.

Commit:
Update quest_greeting_locale RBAC permission ID

- The quest_greeting_locale RBAC permission ID has been increased to 867 from 866.

Commit:
Update and rename 2017_07_30_99_world_335.sql to 2017_10_08_99_world_335.sql

- update the quest_greeting_locale RBAC ID value from 866 to 867

Aokromes committed:
Merge branch '3.3.5' into quest_greetings

Aokromes committed:
Rename 2017_10_08_00_auth.sql to 2017_10_99_99_auth.sql

Aokromes commented:
any feedback on this?

tkrokli commented:
Sorry, I have not had any time to test this or do any kind of research due to real life changes.
I can only hope any other users have got the luxury of time to test & research this topic.

Killyana commented:
Tested and every thing works as intended.

Aokromes committed:
Rename 2017_10_08_99_world_335.sql to 2017_10_29_01_world_335.sql

Aokromes committed:
Rename 2017_10_99_99_auth.sql to 2017_10_29_00_auth.sql

Aokromes committed:
Update world_database.sql

Aokromes committed:
Update world_database.sql

Aokromes committed:
Update auth_database.sql

jackpoz commented:
>9,038 (2,160 direct, 6,878 indirect) bytes in 54 blocks are definitely lost in loss record 57 of 63
   at 0x4C301CA: operator new(unsigned long) (vg_replace_malloc.c:334)
   by 0x240324F: ObjectMgr::LoadQuestGreetings() (ObjectMgr.cpp:6057)
   by 0x2A47D1A: World::SetInitialWorldSettings() (World.cpp:1789)
   by 0x17353DB: main (Main.cpp:246)

@ForesterDev please deallocate any allocated memory on shutdown
